File: java/pgwire/src/test/java/com/risingwave/pgwire/duckdb/JdbcDatabase.java
package com.risingwave.pgwire.duckdb;

import com.risingwave.common.exception.PgErrorCode;
import com.risingwave.common.exception.PgException;
import com.risingwave.pgwire.database.Database;
import com.risingwave.pgwire.database.PgResult;
import java.sql.Connection;
import java.sql.Statement;

public class JdbcDatabase implements Database {
  JdbcDatabase(Connection conn) {
    this.conn = conn;
  }

  @Override
  public PgResult runStatement(String sqlStmt) throws PgException {
    try {
      Statement stmt = conn.createStatement();
      boolean hasResult = stmt.execute(sqlStmt);
      if (hasResult) {
        return new JdbcResult(sqlStmt, stmt.getResultSet());
      }
      return new JdbcResult(sqlStmt, stmt.getUpdateCount());
    } catch (Exception exp) {
      throw new PgException(PgErrorCode.INTERNAL_ERROR, exp);
    }
  }

  @Override
  public String getServerEncoding() {
    return "utf8";
  }

  private final Connection conn;
}

File: legacy/pgwire/src/test/kotlin/com/risingwave/pgwire/duckdb/JdbcDatabase.kt
package com.risingwave.pgwire.duckdb

import com.risingwave.common.exception.PgErrorCode
import com.risingwave.common.exception.PgException
import com.risingwave.pgwire.database.Database
import com.risingwave.pgwire.database.PgResult
import java.lang.Exception
import java.sql.Connection

class JdbcDatabase internal constructor(private val conn: Connection) : Database {
  @Throws(PgException::class)
  override suspend fun runStatement(sqlStmt: String): PgResult {
    return try {
      val stmt = conn.createStatement()
      val hasResult = stmt.execute(sqlStmt)
      if (hasResult) {
        JdbcResult(sqlStmt, stmt.resultSet)
      } else JdbcResult(sqlStmt, stmt.updateCount)
    } catch (exp: Exception) {
      throw PgException(PgErrorCode.INTERNAL_ERROR, exp)
    }
  }

  override val serverEncoding: String
    get() = "utf8"
}
